File: src/lib/ts/helpers/db/pickServerUrl.ts
import { env as envPublic } from '$env/dynamic/public';
import { supabaseAdmin } from '$ts/constants/supabaseAdmin';
import type { TDBServer } from '$ts/types/db';
import type { PostgrestError } from '@supabase/supabase-js';

export async function pickServerUrl(serverUrl: string): Promise<TPickServerUrlRes> {
	if (
		envPublic.PUBLIC_DEFAULT_SERVER_URL === undefined ||
		serverUrl !== envPublic.PUBLIC_DEFAULT_SERVER_URL
	) {
		return { serverUrl: serverUrl, isDefault: false };
	}
	if (!supabaseAdmin) {
		return { serverUrl: serverUrl, isDefault: false };
	}
	const { data, error }: { data: TDBServer[] | null; error: PostgrestError | null } =
		await supabaseAdmin.from('server').select('*').filter('healthy', 'eq', true);
	if (!data || data.length < 1 || error) {
		if (error) console.log(error);
		return { serverUrl: serverUrl, isDefault: false };
	}
	const pickedServer = data[Math.floor(Math.random() * data.length)];
	console.log('---- Picked server url is:', pickedServer.url);
	return { serverUrl: pickedServer.url, isDefault: true };
}

export interface TPickServerUrlRes {
	serverUrl: string;
	isDefault: boolean;
}

File: aux-go/queries/pickServerUrl.go
package queries

import (
	"log"
	"math/rand"
	"time"

	shared "github.com/yekta/stablecog/aux-go/shared"
)

func PickServerUrl(serverUrl string) SServerUrlResult {
	if serverUrl != shared.DEFAULT_SERVER_URL {
		return SServerUrlResult{
			ServerUrl: serverUrl,
			IsDefault: false,
			Error:     false,
		}
	}
	var servers []SDBServer
	_, err := shared.Supabase.From("server").
		Select("id,url,healthy,enabled,created_at,updated_at,last_health_check_at", "", false).
		Filter("enabled", "eq", "true").
		Filter("healthy", "eq", "true").
		ExecuteTo(&servers)
	if err != nil || len(servers) == 0 {
		log.Println(err)
		return SServerUrlResult{
			ServerUrl: serverUrl,
			IsDefault: true,
			Error:     true,
		}
	}
	rand.Seed(time.Now().Unix())
	i := rand.Intn(len(servers))
	pickedServer := servers[i]
	log.Printf("Picked server URL: %v", pickedServer.Url)
	return SServerUrlResult{
		ServerUrl: pickedServer.Url,
		IsDefault: true,
		Error:     false,
	}
}

type SServerUrlResult struct {
	ServerUrl string
	IsDefault bool
	Error     bool
}

type SDBServer struct {
	Id                string `json:"id"`
	Url               string `json:"url"`
	Healthy           bool   `json:"healthy"`
	Enabled           bool   `json:"enabled"`
	CreatedAt         string `json:"created_at"`
	UpdatedAt         string `json:"updated_at"`
	LastHealthCheckAt string `json:"last_health_check_at"`
}
